Wangscape/image/TileRearranger.cpp

```diff
@@ -0,0 +1,263 @@
+    validateColours();
+    // Convert base tile image into stack of partition masks
+    decomposeBaseTile(base_tile);
+    // Check base tile regions are nonempty and nonoverlapping
+    validateBaseTile();
+
+    // Tessellate padded version of base tile in 3x3 grid 
+    const auto tessellation_offsets = makeTessellationOffsets();
+    const auto base_tessellation = tessellated(mTileRearrangement.base, tessellation_offsets);
+    // Check that tessellation has no holes or overlaps
+    validateTessellation(base_tessellation);
+
+    // Find bounding box of dual tile in base tile coordinates
+    calculateRearrangementParameters();
+    // Translate each region of the base tile partition into dual tile partition, and assemble them
+    makeDual();
+
+    // Tessellate padded version of dual tile in 3x3 grid
+    const auto dual_tessellation = tessellated(mTileRearrangement.dual, tessellation_offsets);
+    // Check that tessellation has no holes or overlaps
+    validateTessellation(dual_tessellation);
+
+    // Use adjacency between dual tile mask and offset versions to find edges and corners of dual tile mask
+    findDualBoundaries(dual_tessellation);
+
+    return mTileRearrangement;
+}
+
+void TileRearranger::decomposeBaseTile(const sf::Image & base_tile)
+{
+    const auto base_tile_cube = imageFromSFImage(base_tile);
+    mTileRearrangement.basePartition.set_size(base_tile_cube.n_rows,
+                                               base_tile_cube.n_cols,
+                                               Corners);
+    for (unsigned int c = 0; c < Corners; c++)
+    {
+        auto uniform_colour_cube = ImageColour(arma::size(base_tile_cube));
+        uniform_colour_cube.slice(0).fill(regionColours[c].r);
+        uniform_colour_cube.slice(1).fill(regionColours[c].g);
+        uniform_colour_cube.slice(2).fill(regionColours[c].b);
+        uniform_colour_cube.slice(3).fill(regionColours[c].a);
+        ImageStackGrey channel_matches = arma::conv_to<ImageStackGrey>::from(
+            (base_tile_cube == uniform_colour_cube).eval());
+        // Armadillo doesn't have a product or general reducer function,
+        // so we have to use sum instead.
+        const auto not_function = [](bool x) { return !x; };
+        channel_matches.transform(not_function);
+        ImageGrey pixel_matches = arma::sum(channel_matches, 2);
+        pixel_matches.transform(not_function);
+        mTileRearrangement.basePartition.slice(c) = pixel_matches;
+    }
+    mTileRearrangement.base = arma::sum(mTileRearrangement.basePartition, 2);
+}
+
+void TileRearranger::validateBaseTile() const
+{
+    if (!isNonzero(mTileRearrangement.base))
+    {
+        logError() << "Base tile empty";
+        throw std::runtime_error("Base tile empty");
+    }
+    if (!isBinary(mTileRearrangement.base))
+    {
+        logError() << "Base tile has overlapping regions";
+        throw std::runtime_error("Base tile not correctly partitioned");
```

Comment:
Error messages should be the same I suppose.

Reply:
We have a lot of blocks like this in the code now - one statement writes to `logError()`, the next throws a `std::runtime_error` with a similar or abbreviated message. It might be useful to have a function which can do both of these things, something like:
```cpp
bool x = false;
if(x)
{
    logAndThrowError<std::runtime_error>() << "Logic failure: " << x;
    // creates temporary variable in this block, throws error with the same message when the variable is destructed.
}
```

Reply:
For me better would be treating these functions as assertions and throwing a custom exceptions. The whole part would be just:

```cpp
try
{
    validateRearrangment(mTileRearrangement);
}
catch (const InvalidRearrangementTile& e)
{
    logError() << e.what()
}
```

where `validateRearrangment` calls `assertConnected`, `assertNonZero` and so on. These functions should throw assertion-specific exceptions, like `TileNotConnected`, `EmptyTileRegion`, `IncorrectTilePartition`.